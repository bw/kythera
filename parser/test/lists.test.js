require("jest")
const common = require("./common")

let i = 0

afterEach(() => {
	i += 1
})

describe("Lists", () => {
	let program = common.parseFile("parser/test/lists.ky")
	i = 0

	let t = (name) => {
		test(name, () => expect(program[i]).toMatchSnapshot())
	}

	t("Initialization with 'let'")
	t("Access")
	t("Creation of custom type (irrelvant)")
	// i += 1
	t("Initialization with custom type")
	t("Initialization from object list literal")
	t("Function returning list")
	t("Access from function call result")
})