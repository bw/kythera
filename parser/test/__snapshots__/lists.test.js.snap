// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lists Access 1`] = `
ParseNode {
  "index": ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
    "value": 2,
  },
  "kind": "access",
  "target": ParseNode {
    "kind": "identifier",
    "name": "myList",
  },
  "type": "unknown",
}
`;

exports[`Lists Access from function call result 1`] = `
ParseNode {
  "index": ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
    "value": 10,
  },
  "kind": "access",
  "target": ParseNode {
    "arguments": Array [],
    "kind": "call",
    "target": ParseNode {
      "kind": "identifier",
      "name": "myListFn",
    },
  },
  "type": "unknown",
}
`;

exports[`Lists Creation of custom type (irrelvant) 1`] = `
ParseNode {
  "identifier": "myType",
  "kind": "let",
  "value": ParseNode {
    "kind": "literal",
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "type",
    },
    "value": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "structure": Object {
        "a": ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "int",
        },
        "b": ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "str",
        },
      },
      "type": "obj",
    },
  },
}
`;

exports[`Lists Function returning list 1`] = `
ParseNode {
  "identifier": "myListFn",
  "kind": "let",
  "value": ParseNode {
    "body": Array [
      ParseNode {
        "kind": "return",
        "value": ParseNode {
          "kind": "new",
          "target": ParseNode {
            "contains": ParseNode {
              "kind": "type",
              "origin": "builtin",
              "type": "int",
            },
            "kind": "type",
            "origin": "builtin",
            "type": "list",
          },
        },
      },
    ],
    "kind": "literal",
    "parameters": Array [],
    "returns": ParseNode {
      "contains": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "kind": "type",
      "origin": "builtin",
      "type": "list",
    },
    "type": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "parameters": Array [],
      "returns": ParseNode {
        "contains": ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "int",
        },
        "kind": "type",
        "origin": "builtin",
        "type": "list",
      },
      "type": "fn",
    },
  },
}
`;

exports[`Lists Initialization from object list literal 1`] = `
ParseNode {
  "identifier": "myTypeList2",
  "kind": "let",
  "value": ParseNode {
    "kind": "new",
    "target": ParseNode {
      "contains": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "structure": Object {
          "a": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
          "b": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "str",
          },
        },
        "type": "obj",
      },
      "kind": "type",
      "origin": "builtin",
      "type": "list",
    },
  },
}
`;

exports[`Lists Initialization with 'let' 1`] = `
ParseNode {
  "identifier": "myList",
  "kind": "let",
  "value": ParseNode {
    "kind": "new",
    "target": ParseNode {
      "contains": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "kind": "type",
      "origin": "builtin",
      "type": "list",
    },
  },
}
`;

exports[`Lists Initialization with custom type 1`] = `
ParseNode {
  "identifier": "myTypeList",
  "kind": "let",
  "value": ParseNode {
    "kind": "new",
    "target": ParseNode {
      "contains": ParseNode {
        "kind": "type",
        "name": "myType",
        "origin": "named",
      },
      "kind": "type",
      "origin": "builtin",
      "type": "list",
    },
  },
}
`;

exports[`Lists List literals bool 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": true,
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "bool",
      },
      "value": false,
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "bool",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals float 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
      "value": 1.234,
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
      "value": 3.14159,
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "float",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals fn 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "body": Array [
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "kind": "type",
              "origin": "builtin",
              "type": "null",
            },
            "value": null,
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [
        Object {
          "name": "a",
          "type": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
        },
      ],
      "returns": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "null",
      },
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
        ],
        "returns": ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "null",
        },
        "type": "fn",
      },
    },
    ParseNode {
      "body": Array [
        ParseNode {
          "kind": "return",
          "value": ParseNode {
            "kind": "literal",
            "type": ParseNode {
              "kind": "type",
              "origin": "builtin",
              "type": "null",
            },
            "value": null,
          },
        },
      ],
      "kind": "literal",
      "parameters": Array [
        Object {
          "name": "b",
          "type": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
        },
      ],
      "returns": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "null",
      },
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
        ],
        "returns": ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "null",
        },
        "type": "fn",
      },
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "parameters": Array [
        ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "int",
        },
      ],
      "returns": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "null",
      },
      "type": "fn",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals int 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 1,
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 2,
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 3,
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
      "value": 4,
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "int",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals null 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "null",
      },
      "value": null,
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "null",
      },
      "value": null,
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "null",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals obj 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "structure": Object {},
        "type": "obj",
      },
      "value": Object {
        "a": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
          "value": 0,
        },
        "b": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "str",
          },
          "value": "asdf",
        },
      },
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "structure": Object {},
        "type": "obj",
      },
      "value": Object {
        "a": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
          "value": 3,
        },
        "b": ParseNode {
          "kind": "literal",
          "type": ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "str",
          },
          "value": "qwerty",
        },
      },
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "structure": Object {},
      "type": "obj",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals str 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "str",
      },
      "value": "hello",
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "str",
      },
      "value": "world",
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "str",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;

exports[`Lists List literals type 1`] = `
ParseNode {
  "elements": Array [
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "type",
      },
      "value": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "int",
      },
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "type",
      },
      "value": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "float",
      },
    },
    ParseNode {
      "kind": "literal",
      "type": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "type": "type",
      },
      "value": ParseNode {
        "kind": "type",
        "origin": "builtin",
        "parameters": Array [
          ParseNode {
            "kind": "type",
            "origin": "builtin",
            "type": "int",
          },
        ],
        "returns": ParseNode {
          "kind": "type",
          "origin": "builtin",
          "type": "null",
        },
        "type": "fn",
      },
    },
  ],
  "kind": "literal",
  "type": ParseNode {
    "contains": ParseNode {
      "kind": "type",
      "origin": "builtin",
      "type": "type",
    },
    "kind": "type",
    "origin": "builtin",
    "type": "list",
  },
}
`;
